### ISSUE TITLE
Named services

### ISSUE THREAD

**Issue by jone** [plone/plone.rest#16]
For being able to extend the API, I think we need to support named services.

Some use cases:
- Searching content, eg: `GET /search?portal_type=Folder`
- Managing users (list, add, update), eg: `GET /users/john.doe`
- Change workflow state: `PUT /document-1/workflow`
- Versions (CMFEditions): `GET /document-1/versions`, `GET /document-1/versions/3`

It would be awesome if it is possible to nest named services too.

We have tried to implement that by just adding the name to the `IBrowserPublisher` service adapter, but the traversal then does not work properly:
```python
Traceback (innermost last):
Module ZPublisher.Publish, line 249, in publish_module_standard
Module ZPublisher.Publish, line 177, in publish
Module ZPublisher.Publish, line 127, in publish
Module ZPublisher.BaseRequest, line 556, in traverse
AttributeError: namedservice
```

I've pushed my experiment to the `jone-named-service` branch: https://github.com/plone/plone.rest/compare/jone-named-service

What do you think in general about supporting named services?
Any hints how we can get that working?

**Comment by tisto** [comment]
We discussed named services but decided to focus on the REST methods on existing content first. It would be highly appreciated if you would look into this! I would just like to avoid adding an additional traversal mechanism like in plone.jsonapi.

**Comment by jone** [comment]
Seems related to https://github.com/plone/plone.restapi/issues/21

I agree that we should reimplement traversal but use the existing Zope traversal. I think we just need to get around the Zope traversal so that it accepts named services.

**Comment by tisto** [comment]
Sounds good!

**Comment by lukasgraf** [comment]
The issue with the branch referenced by @jone is that named services currently fail because of the [special WebDav handling in `BaseRequest.traverse()](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/BaseRequest.py#L538-L556): Because `request.maybe_webdav_client == 1` the `no_acquire_flag` always get's set, and [this check happens quite early](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/BaseRequest.py#L389-L400), namely before any [`__before_publishing_traverse__'` hooks](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/BaseRequest.py#L442-L444) are called that could influence that.

I therefore tried setting `request.maybe_webdav_client = 0` in a `PubStart` event, which seems to work: 

https://github.com/plone/plone.rest/compare/lg-named-service

After that, I ran into the issue that `Unauthorized` was raised when traversing to named services. The security check failed somewhere deep in [`ZopeSecurityPolicy.validate()`](https://github.com/zopefoundation/AccessControl/blob/master/src/AccessControl/ImplPython.py#L225-L451). Based on a hunch, I let the `Service` class inherit from `Acquisition.Implicit`, and voilà, that seems to do the trick ;-) We certainly need to look into this some more, especially the security validation, but it does look more promising than this afternoon.

/cc @buchi

**Comment by buchi** [comment]
Fixed in #18
